### PR TITLE
NPE in ApiBaseline on adding new JVM

### PR DESCRIPTION
**Pull request by iloveeclipse**
An attempt to reduce the probability of errors while API baseline is being updated due the JVM change.

- Move rebindVM() to a job and let it wait for pending API analysis jobs
- changed most of internal maps/lists to concurrent versions
- made initialize() synchronized

Fixes https://github.com/eclipse-pde/eclipse.pde/issues/1042

### CODE REVIEW

File: apitools/org.eclipse.pde.api.tools/src/org/eclipse/pde/api/tools/internal/model/ApiBaseline.java [hunk 1]
```diff
@@ -966,36 +947,37 @@ public void vmChanged(PropertyChangeEvent event) {
...
+				return;
+			}
+			fVMBinding = null;
+			IApiComponent[] components = getApiComponents();
+			HashSet<String> ees = new HashSet<>();
+			for (IApiComponent component2 : components) {
+				ees.addAll(Arrays.asList(component2.getExecutionEnvironments()));
+			}
+			resolveSystemLibrary(ees);
+		}).schedule();
```
Comment (laeubi) [inline apitools/org.eclipse.pde.api.tools/src/org/eclipse/pde/api/tools/internal/model/ApiBaseline.java:974]: should one probably `join()` on the job here? Previously the caller could be sure that the method does not return before the rebind was done.

Comment (iloveeclipse) [reply apitools/org.eclipse.pde.api.tools/src/org/eclipse/pde/api/tools/internal/model/ApiBaseline.java:974]: The only caller was a listener. For the listener on JRE changes it is irrelevant to wait on this code to complete.

The reason to use a job here was the possibility of a deadlocks here, after initialize() was made synchronized.

File: apitools/org.eclipse.pde.api.tools/src/org/eclipse/pde/api/tools/internal/model/ApiBaseline.java [hunk 2]
```diff
@@ -177,6 +179,10 @@ public class ApiBaseline extends ApiElement implements IApiBaseline, IVMInstallC
 	public ApiBaseline(String name) {
 		super(null, IApiElement.BASELINE, name);
 		fComponentsProvidingPackageCache = new ConcurrentHashMap<>(8);
+		fSystemLibraryComponentList = new CopyOnWriteArrayList<>();
+		fComponentsById = new ConcurrentHashMap<>();
+		fAllComponentsById = new ConcurrentHashMap<>();
+		fComponentsByProjectNames = new ConcurrentHashMap<>();
```
Comment (HannesWell) [inline apitools/org.eclipse.pde.api.tools/src/org/eclipse/pde/api/tools/internal/model/ApiBaseline.java:185]: You could already initialize these fields at their declaration. Would save the lines here and easier to determine what implementation is actually used.

Comment (iloveeclipse) [reply apitools/org.eclipse.pde.api.tools/src/org/eclipse/pde/api/tools/internal/model/ApiBaseline.java:185]: I prefer to init in constructor, this saves you many jumps while debugging.

**Comment by iloveeclipse** [comment]
There are no test fails, despite the "quality gate" reporting the opposite:

image

and the only "errors" are **unrelated** messages from maven:
```
[INFO] --- tycho-eclipse:4.0.5-SNAPSHOT:eclipse-build (generate-xmls) @ org.eclipse.pde.ds.tck ---
[ERROR] tck/OSGI-INF/impl-src/org/osgi/impl/bundle/component/annotations/HelloWorld10.java:27 Invalid component descriptor namespace: http://www.osgi.org/xmlns/scr/v1.0.0
[ERROR] tck/OSGI-INF/impl-src/org/osgi/impl/bundle/component/annotations/ServiceBundle.java:30 Property 'servicefactory' is ignored when non-default scope is specified.
[ERROR] tck/OSGI-INF/impl-src/org/osgi/impl/bundle/component/annotations/ServiceSingleton.java:30 Property 'servicefactory' is ignored when non-default scope is specified.
[INFO]
```

**Pull request merged**